src/main/java/org/folio/auth/authtokenmodule/tokens/Token.java

```diff
@@ -0,0 +1,248 @@
+    String decodedJson = new String(Base64.getDecoder().decode(encodedJson));
+    return new JsonObject(decodedJson);
+  }
+
+  /**
+   * Returns true if the token is encrypted, otherwise false.
+   * @param token A string version of the token to test.
+   * @return True if the token is encrypted, otherwise false.
+   * @throws TokenValidationException
+   */
+  public static boolean isEncrypted(String token) throws TokenValidationException {
+    // This is based on how JOSEObject parses encrypted tokens. It checks the number of "parts"
+    // which are equivalent to the number of "." separators in the token. If there are 5 parts
+    // it is considered to be encrypted. If there are not 5 it is rejected from JWE parsing.
+    // Our unencrypted tokens have 3 parts (and 2 separators).
+    int parts = token.split("\\.").length;
+    if (parts == 5)
+      return true;
+
+    if (parts == 3)
+      return false;
+    
+    throw new TokenValidationException("Unexpected token part count", 401);
+  }
+
+  /**
+   * Validate all the things that tokens have in common.
+   * @param request The http request context where the token is being provided.
+   * @throws TokenValidationException Throws this when any validation step fails, but may throw
+   * other exceptions as well.
+   */
+  protected void validateCommon(HttpServerRequest request) throws TokenValidationException {
+    try {
+      // Check that the token has a source.
+      if (source == null)
+        throw new TokenValidationException("Token has no source defined", 500);
+
+      // Check that the claims have been created.
+      if (claims == null)
+        throw new TokenValidationException("Token has no claims", 500);
+
+      // Check some claims that all tokens must have.
+      String[] requiredClaims = new String[] { "sub", "tenant", "type" };
+      for (String c : requiredClaims) {
+        if (!claims.containsKey(c)) {
+          throw new TokenValidationException(String.format("Token is missing %s claim", c), 500);
+        }
+      }
+
+      if (request == null)
+        return;
+
+      // Check that some items in the headers match what are in the token.
+      String headerTenant = request.headers().get(XOkapiHeaders.TENANT);
+      if (!claims.getString("tenant").equals(headerTenant))
+        throw new TokenValidationException("Tenant in header does not equal tenant in token", 403);
+
+      String headerUserId = request.headers().get(XOkapiHeaders.USER_ID);
+      if (headerUserId != null && !claims.getString("user_id").equals(headerUserId))
+        throw new TokenValidationException("User id in header does not equal user id in token", 403);
+
+    } catch (TokenValidationException e) {
+      throw e;
+    } catch (Exception e) {
+      throw new TokenValidationException("Unexpected token validation exception", e, 500);
+    }
+  }
+
+  private static Token parse(String sourceToken, TokenCreator tokenCreator) throws TokenValidationException {
```

Comment:
Can you move this method into a `TokenFactory` class and rename it to `create`?

Reply:
I had that originally since it is the more canonical abstract factory pattern but then I thought it might be better to make this method private since validation is the only case when tokens will need to be created with these constructors. In other words there's no case when the factory would need to be used and we don't want to encourage it. Let me know what you think.